src/Parser.js

```diff
@@ -371,10 +403,54 @@ Parser.prototype.parseImplicitGroup = function(pos, mode) {
         } else {
             throw new ParseError("Missing \\right", this.lexer, body.position);
         }
+    } else if (func === "\\begin") {
+        // begin...end is similar to left...right
+        var begin = this.parseFunction(pos, mode);
+        var envName = begin.result.value.name;
+        if (!environments.hasOwnProperty(envName)) {
+            throw new ParseError(
+                "No such environment: " + envName,
+                this.lexer, begin.result.value.namepos);
+        }
+        // Build the environment object. Arguments and other information will
+        // be made available to the begin and end methods using properties.
+        var env = new environments[envName](envName, this);
+        env.parser = this;
+        env.lexer = this.lexer;
+        env.mode = mode;
+        env.beginNode = begin;
+        env.args = [];
```

Comment:
Not needed by most environments, but offered to them. The most important of these are `args`, `argPositions` and of course `body`. These can't be passed to the constructor. But even if they could, I'd rather not do this. There are potentially many environments, but there is only one place where instances of these are created. So having that code just in this one place avoids code duplication. Particularly since environments, the way I implemented them, don't call a base class constructor for common functionality.